rmw_zenoh_cpp/src/detail/graph_cache.cpp

```diff
@@ -41,170 +41,284 @@ using Entity = liveliness::Entity;
-        return;
-      }
+  // First update the topic map within the node.
+  if (entity.type() == EntityType::Publisher) {
+    update_topic_map_for_put(graph_node->pubs_, entity);
+  } else if (entity.type() == EntityType::Subscription) {
+    update_topic_map_for_put(graph_node->subs_, entity);
+  } else if (entity.type() == EntityType::Service) {
+    update_topic_map_for_put(graph_node->services_, entity);
+  } else {
+    update_topic_map_for_put(graph_node->clients_, entity);
+  }
 
-      const liveliness::TopicInfo topic_info = entity.topic_info().value();
-      std::string entity_desc = "";
-      GraphNode::TopicMap & topic_map =
-        [](const Entity & entity, GraphNode & graph_node,
-          std::string & entity_desc) -> GraphNode::TopicMap &
-        {
-          if (entity.type() == EntityType::Publisher) {
-            entity_desc = "publisher";
-            return graph_node.pubs_;
-          } else if (entity.type() == EntityType::Subscription) {
-            entity_desc = "subscription";
-            return graph_node.subs_;
-          } else if (entity.type() == EntityType::Service) {
-            entity_desc = "service";
-            return graph_node.services_;
+  // Then update the variables tracking topics across the graph.
+  // We invoke update_topic_map_for_put() with report_events set to true for
+  // pub/sub.
+  if (entity.type() == EntityType::Publisher ||
+    entity.type() == EntityType::Subscription)
+  {
+    update_topic_map_for_put(this->graph_topics_, entity, true);
+  } else {
+    update_topic_map_for_put(this->graph_services_, entity);
+  }
+}
 
+///=============================================================================
+void GraphCache::update_topic_map_for_put(
+  GraphNode::TopicMap & topic_map,
+  const liveliness::Entity & entity,
+  bool report_events)
+{
+  if (!entity.topic_info().has_value()) {
+    // This should not happen as topic_info should be populated for all non-node entites.
+    RCUTILS_LOG_WARN_NAMED(
+      "rmw_zenoh_cpp",
+      "update_topic_map_for_put() called for non-node entity without valid TopicInfo. "
+      "Report this.");
+    return;
+  }
+  const liveliness::TopicInfo topic_info = entity.topic_info().value();
+
+  // For the sake of reusing data structures and lookup functions, we treat publishers and
+  // clients as equivalent. Similarly, subscriptions and services are equivalent.
+  const bool is_pub = is_entity_pub(entity);
+  // Initialize a map that will be populated with any QoS events that may be detected.
+  std::unordered_map<liveliness::Entity, std::unordered_set<rmw_zenoh_event_type_t>>
+  local_entities_with_events = {};
+  std::unordered_set<liveliness::Entity> pubs = {};
+  std::unordered_set<liveliness::Entity> subs = {};
+  if (is_pub) {
+    pubs.insert(entity);
+  } else {
+    subs.insert(entity);
+  }
+  TopicDataPtr graph_topic_data = std::make_shared<TopicData>(
+    topic_info,
+    std::move(pubs),
+    std::move(subs));
```

Comment:
Actually, this relates to my previous comment.  It might make more sense to pass the `entity` into the `TopicData` constructor, and then just have `TopicData` own the `pubs` and `subs` unordered_sets.  That would resolve the ownership question.

Reply:
Hoping this makes things a bit cleaner https://github.com/ros2/rmw_zenoh/pull/103/commits/00e8e1c2929849e33913161b6d442e55050457d7